File: ui/src/api/days.ts
import { apiGet } from './client'
import type { PaginatedResponse, DayListItem, DayDetail } from '@/types'

export interface DaysParams {
    limit?: number
    offset?: number
    from_date?: string
    to_date?: string
    device_id?: number
}

export const getDays = apiGet<PaginatedResponse<DayListItem>, [params?: DaysParams]>(
    '/days/',
    (params = {}) => ({ params }),
)

export const getDay = apiGet<DayDetail, [date: string]>((date) => `/days/${date}`)

export const getDates = apiGet<{ dates: string[] }>('/days/dates')

File: src/snore/api/routers/days.py
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Depends, Query

from snore.api.deps import DateRangeParams, PaginationParams, service_dep
from snore.api.schemas import (
    DateListResponse,
    DayDetail,
    DayListItem,
    PaginatedResponse,
)
from snore.services import DayService

router = APIRouter()

DayServiceDep = Annotated[DayService, Depends(service_dep(DayService))]


@router.get("/", response_model=PaginatedResponse[DayListItem])
async def list_days(
    service: DayServiceDep,
    pagination: PaginationParams = Depends(),
    dates: DateRangeParams = Depends(),
    device_id: int | None = Query(default=None),
) -> PaginatedResponse[DayListItem]:
    items, total = await service.list_days(
        from_date=dates.from_date,
        to_date=dates.to_date,
        device_id=device_id,
        limit=pagination.limit,
        offset=pagination.offset,
    )
    return PaginatedResponse(
        items=items, total=total, limit=pagination.limit, offset=pagination.offset
    )


@router.get("/dates", response_model=DateListResponse)
async def list_dates(service: DayServiceDep) -> DateListResponse:
    dates = await service.list_dates()
    return DateListResponse(dates=dates)


@router.get("/{day_date}", response_model=DayDetail)
async def get_day(day_date: date, service: DayServiceDep) -> DayDetail:
    return await service.get_day(day_date)
